src/record/record.js

```diff
@@ -345,10 +324,13 @@ Record.prototype._recoverRecord = function( remoteVersion, remoteData, message )
  * @private
  * @returns {void}
  */
-Record.prototype._onRecordRecovered = function( remoteVersion, error, data ) {
+Record.prototype._onRecordRecovered = function( remoteVersion, remoteData, error, data ) {
 	if( !error ) {
 		this.version = remoteVersion;
-		this.set( data );
+		if ( data !== this._$data && !utils.deepEquals( data, this._$data ) ) {
```

Comment:
Gotcha, calling set traps the local incorrect state.

On a side note, this PR is getting huge. I'm not against most of the changes but be great to not do any actual functionality changes as it's going to be hard enough to test!